oauth2_http/java/com/google/auth/oauth2/TokenVerifier.java

```diff
@@ -0,0 +1,405 @@
+import com.google.common.base.Preconditions;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.UncheckedExecutionException;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.math.BigInteger;
+import java.security.AlgorithmParameters;
+import java.security.GeneralSecurityException;
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.PublicKey;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.spec.ECGenParameterSpec;
+import java.security.spec.ECParameterSpec;
+import java.security.spec.ECPoint;
+import java.security.spec.ECPublicKeySpec;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.InvalidParameterSpecException;
+import java.security.spec.RSAPublicKeySpec;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Handle verification of Google-signed JWT tokens.
+ *
+ * @author Jeff Ching
+ * @since 0.21.0
+ */
+@Beta
+public class TokenVerifier {
+  private static final String IAP_CERT_URL = "https://www.gstatic.com/iap/verify/public_key-jwk";
+  private static final String FEDERATED_SIGNON_CERT_URL =
+      "https://www.googleapis.com/oauth2/v3/certs";
+  private static final Set<String> SUPPORTED_ALGORITHMS = ImmutableSet.of("RS256", "ES256");
+
+  private final String audience;
+  private final String certificatesLocation;
+  private final String issuer;
+  private final PublicKey publicKey;
+  private final Clock clock;
+  private final LoadingCache<String, Map<String, PublicKey>> publicKeyCache;
+
+  /**
+   * Construct the default TokenVerifier. By default, this verifier only validates the signature.
+   */
+  public TokenVerifier() {
+    this(newBuilder());
+  }
+
+  private TokenVerifier(Builder builder) {
+    this.audience = builder.audience;
+    this.certificatesLocation = builder.certificatesLocation;
+    this.issuer = builder.issuer;
+    this.publicKey = builder.publicKey;
+    this.clock = builder.clock;
+    this.publicKeyCache =
+        CacheBuilder.newBuilder()
+            .expireAfterWrite(1, TimeUnit.HOURS)
+            .build(new PublicKeyLoader(builder.httpTransportFactory));
+  }
+
+  public static Builder newBuilder() {
+    return new Builder()
+        .setClock(Clock.SYSTEM)
+        .setHttpTransportFactory(OAuth2Utils.HTTP_TRANSPORT_FACTORY);
+  }
+
+  /**
+   * Verify an encoded JWT token.
+   *
+   * @param token encoded JWT token
+   * @return The parsed JsonWebSignature instance for additional validation if necessary
```

Comment:
Please check the doc comments to make sure they follow the Oracle guidelines: no initial cap and no period unless there's a complete sentence